### PR TITLE
Metadata cache by topic id and fixes

### PR DESCRIPTION
**Pull request by emasab**
Metadata cache by topic id is added, with an AVL tree having a Uuid as key and the same struct as value.

### General fixes

 * Metadata cache was cleared on full metadata refresh, leading to unnecessary
   refreshes and occasional `UNKNOWN_TOPIC_OR_PART` errors. Solved by updating
   cache for existing or hinted entries instead of clearing them.
   Happening since 2.1.0.
 * A metadata call before member joins consumer group,
   could lead to an `UNKNOWN_TOPIC_OR_PART` error. Solved by updating
   the consumer group following a metadata refresh only in safe states.
   Happening since 2.1.0.
 * Metadata refreshes without partition leader change could lead to a loop of
   metadata calls at fixed intervals. Solved by stopping metadata refresh when
   all existing metadata is non-stale. Happening since 2.3.0.
 * A partition migration could happen, using stale metadata, when the partition
   was undergoing a validation and being retried because of an error.
   Solved by doing a partition migration only with a non-stale leader epoch.
   Happening since 2.1.0.

### CODE REVIEW

**Comment by emasab** [comment]
Closes #4577

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,36 @@ librdkafka v2.3.1 is a maintenance release:
...
+ * Fix to main loop timeout calculation leading to a tight loop for a period
+   of 1 ms max (#4660).
+
+
+## Fixes
+
+### General fixes
+
+ * Metadata cache was cleared on full metadata refresh, leading to unnecessary
+   refreshes and accasional `UNKNOWN_TOPIC_OR_PART` errors. Solved by updating
```
Comment (Quuxplusone) [inline CHANGELOG.md:23]:
```suggestion
   refreshes and occasional `UNKNOWN_TOPIC_OR_PART` errors. Solved by updating
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,36 @@ librdkafka v2.3.1 is a maintenance release:
...
+   cache for existing or hinted entries instead of clearing them.
+   Happening since 2.1.0 (#4660).
+ * Metadata refreshes without partition leader change could lead to a loop of
+   intervaled metadata calls. Solved by stopping metadata refresh when
```
Comment (Quuxplusone) [inline CHANGELOG.md:27]: FWIW, "intervaled" isn't a word; I don't have enough context to know what this means.

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,36 @@ librdkafka v2.3.1 is a maintenance release:
...
+   all existing metadata is non-stale. Happening since 2.3.0 (#4660).
+ * A partition migration could happen, using stale metadata, when the partition
+   was undergoing a validation and being retried because of an error.
+   Solved by doing a partition migration only with a non-stale leader epoch.
+   Happening since 2.1.0 (#4660).
+ * In librdkafka main thread loop, when it was awaken less that 1 ms
```
Comment (Quuxplusone) [inline CHANGELOG.md:33]:
```suggestion
 * When the main thread loop was awakened less than 1 ms
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,36 @@ librdkafka v2.3.1 is a maintenance release:
...
+   before the expiration of a timeout, it was serving with a zero timeout,
+   leading to increased CPU usage until the timeout was reached.
+   Happening since 1.x (#4660).
```
Comment (Quuxplusone) [inline CHANGELOG.md:36]: FWIW, it's mildly confusing that all these bullets point to #4660 (this PR), instead of to the actual bug reports about the bugs being described. Do such bug reports exist?

File: src/rdkafka.c [hunk 5]
```diff
@@ -2127,7 +2127,12 @@ static int rd_kafka_thread_main(void *arg) {
                                        RD_KAFKA_CGRP_STATE_TERM)))) {
                 rd_ts_t sleeptime = rd_kafka_timers_next(
                     &rk->rk_timers, 1000 * 1000 /*1s*/, 1 /*lock*/);
-                rd_kafka_q_serve(rk->rk_ops, (int)(sleeptime / 1000), 0,
+                /* Round up to avoid calling serve with timeout 0 ms
+                 * in a tight loop until 1 ms has passed. */
+                int timeout_ms = (int)(sleeptime / 1000);
+                if (sleeptime % 1000 > 0)
+                        timeout_ms++;
```
Comment (Quuxplusone) [inline src/rdkafka.c:2134]: This is the same as
```
int timeout_ms = (int)(sleeptime / 1000) + (sleeptime % 1000 > 0);
```
or perhaps even better,
```
int timeout_ms = ((sleeptime + 999) / 1000);
```
(The cast to `int` probably isn't needed, unless it's suppressing a compiler warning.)

Also, this diff seems totally unrelated to the "metadata cache" stuff in this PR. This should be a separate commit with a separate PR.

File: src/rdkafka_metadata_cache.c [hunk 6]
```diff
@@ -221,6 +209,25 @@ rd_kafka_metadata_cache_find(rd_kafka_t *rk, const char *topic, int valid) {
...
+/**
+ * @brief Find cache entry by topic id
+ *
+ * @param valid: entry must be valid (not hint)
+ *
+ * @locks rd_kafka_*lock()
+ */
+struct rd_kafka_metadata_cache_entry *
+rd_kafka_metadata_cache_find_by_id(rd_kafka_t *rk,
+                                   const rd_kafka_Uuid_t topic_id,
```
Comment (Quuxplusone) [inline src/rdkafka_metadata_cache.c:221]: `const rd_kafka_Uuid_t` passes "by const value," which is never correct. Surely the compiler warned about this?

Everywhere else in librdkafka that takes `rd_kafka_Uuid_t` either takes it by pointer to avoid a copy — `const rd_kafka_Uuid_t*` — which is probably what you want here, since you never use the parameter's own value and it's large — or takes it by "non-const value" in two places (`rd_kafka_topic_partition_new_with_topic_id` and `rd_kafka_Uuid_cmp`: the latter is inline and the former looks like a flaw to me but it's part of the public API now).

File: src/rdkafka_mock.c [hunk 7]
```diff
@@ -469,7 +469,38 @@ rd_kafka_mock_partition_assign_replicas(rd_kafka_mock_partition_t *mpart,
...
+        rd_kafka_mock_partition_leader_t *leader_response;
 
+        leader_response               = rd_calloc(1, sizeof(*leader_response));
+        leader_response->leader_id    = leader_id;
+        leader_response->leader_epoch = leader_epoch;
+        TAILQ_INSERT_TAIL(&mpart->leader_responses, leader_response, link);
+}
+
+/**
+ * @brief First forced partition leader response in passed \p mpart .
```
Comment (Quuxplusone) [inline src/rdkafka_mock.c:488]: FWIW I don't understand this grammar. Does it mean `\p mpart holds the first forced response from the partition leader`, or I guess more likely `Fetch the first partition leader response into \p mpart`?

File: CHANGELOG.md (hunk 1 above)
Comment (emasab) [reply CHANGELOG.md]: Thanks, fixed it

File: CHANGELOG.md (hunk 2 above)
Comment (emasab) [reply CHANGELOG.md:31]: intervaled means at fixed intervals, it's rarely used I think. I can put the phrase "at fixed intervals" instead

File: src/rdkafka.c (hunk 5 above)
Comment (emasab) [reply src/rdkafka.c:2134]: Thanks, I think I'll use the second one

File: CHANGELOG.md (hunk 4 above)
Comment (emasab) [reply CHANGELOG.md:40]: We usually put the id of the PR containing the changes there, then the PR needs to link to the corresponding issues that closes. I update this PR

File: src/rdkafka_metadata_cache.c (hunk 6 above)
Comment (emasab) [reply src/rdkafka_metadata_cache.c:221]: Internally we pass the Uuid by struct copy, as a primitive value, to avoid more heap allocations. Here the const isn't strictly necessary, but it doesn't hurt as that parameter shouldn't be modified inside the function

File: src/rdkafka.c (hunk 5 above)
Comment (emasab) [reply src/rdkafka.c:2134]: Putting it in a different commit too

File: src/rdkafka_metadata.c [hunk 8]
```diff
@@ -954,17 +1001,17 @@ rd_kafka_parse_Metadata0(rd_kafka_broker_t *rkb,
...
-                rd_kafka_metadata_cache_expiry_start(rk);
         }
-
-
-        // TODO: Should be done for requested_topic_ids as well.
         /* Remove cache hints for the originally requested topics. */
         if (requested_topics)
                 rd_kafka_metadata_cache_purge_hints(rk, requested_topics);
+        if (requested_topic_ids)
+                rd_kafka_metadata_cache_purge_hints(rk, requested_topic_ids);
```
Comment (pranavrth) [inline src/rdkafka_metadata.c:1009]: `rd_kafka_metadata_cache_purge_hints_by_id`

File: src/rdkafka_metadata.c [hunk 9]
```diff
@@ -954,17 +1001,17 @@ rd_kafka_parse_Metadata0(rd_kafka_broker_t *rkb,
...
+
+        if (cache_changes) {
+                rd_kafka_metadata_cache_propagate_changes(rk);
+                rd_kafka_metadata_cache_expiry_start(rk);
```
Comment (pranavrth) [inline src/rdkafka_metadata.c:1013]: - Do you mean to include expiry start call inside the if condition? Earlier it was outside.
- What happens to the cache entries which were not updated? Do we restart the timer for them as well?

**Review by pranavrth — CHANGES_REQUESTED** [review]
Reviewed few files which I understood. Reviewing more. Please check comments in the original PR as well.

File: src/rdkafka_mock.c (hunk 7 above)
Comment (emasab) [reply src/rdkafka_mock.c:488]: It's:
`Return the first forced partition leader response in \p mpart , if available`

File: src/rdkafka_metadata.c [hunk 10]
```diff
@@ -474,12 +474,14 @@ rd_kafka_parse_Metadata0(rd_kafka_broker_t *rkb,
...
-
-        const rd_list_t *requested_topics = request_topics;
-        rd_bool_t all_topics              = rd_false;
-        rd_bool_t cgrp_update             = rd_false;
+        const int log_decode_errors  = LOG_ERR;
+        rd_list_t *missing_topics    = NULL;
+        rd_list_t *missing_topic_ids = NULL;
+
+        const rd_list_t *requested_topics    = request_topics;
+        const rd_list_t *requested_topic_ids = NULL;
```
[the remark this answers is not in the document]
Comment (emasab) [reply src/rdkafka_metadata.c:482]: Updated the documentation for the three versions of MetadataRequests

File: src/rdkafka_metadata.c (hunk 9 above)
Comment (emasab) [reply src/rdkafka_metadata.c:1016]: Here it's restarting the timer if some entries have been updated, the timer is calculated from the expiry time of the first element, the one that expires first. Earlier was outside just in this code, in trunk it's done only if `!all_topics` but now existing entries can be updated in that case.

**Review by pranavrth — CHANGES_REQUESTED** [review]
Some review for Cache.

File: src/rdkafka_metadata.c [hunk 11]
```diff
@@ -831,34 +839,37 @@ rd_kafka_parse_Metadata0(rd_kafka_broker_t *rkb,
...
+                if (requested_topic_ids)
+                        rd_list_free_cb(
+                            missing_topic_ids,
+                            rd_list_remove_cmp(missing_topic_ids,
+                                               &mdi->topics[i].topic_id,
+                                               (void *)rd_kafka_Uuid_ptr_cmp));
+                /* Only update cache when not asking
+                 * for all topics or cache entry
+                 * already exists. */
+                rd_kafka_wrlock(rk);
```
[the remark this answers is not in the document]
Comment (emasab) [reply src/rdkafka_metadata.c:856]: To call `rd_kafka_metadata_cache_find `and `rd_kafka_metadata_cache_find_by_id` the lock is required because it's reading the AVL tree, it could use a read lock but that cannot be upgraded to a write lock and if releasing it, after acquiring the write lock it should have to check again if the entry was deleted in the meantime.
I think the lock is rarely contended here and in that case the futex implementation will ensure no system calls are made.

File: src/rdkafka_metadata.c [hunk 12]
```diff
@@ -890,6 +901,41 @@ rd_kafka_parse_Metadata0(rd_kafka_broker_t *rkb,
...
+                        rd_kafka_rdunlock(rk);
+                        if (rkt) {
+                                /* Received metadata response contained no
+                                 * information about topic 'rkt' and thus
+                                 * indicates the topic is not available in the
+                                 *  cluster.
+                                 * Mark the topic as non-existent */
+                                rd_kafka_topic_wrlock(rkt);
+                                rd_kafka_topic_set_notexists(
+                                    rkt, RD_KAFKA_RESP_ERR__UNKNOWN_TOPIC);
```
[the remark this answers is not in the document]
Comment (emasab) [reply src/rdkafka_metadata.c:934]: This is the local error `_UNKNOWN_TOPIC`, not the broker one `UNKNOWN_TOPIC_OR_PART` or `UNKNOWN_TOPIC_ID` so it's correct to use the same code to signal that the topic is unknown, independently of which parameter was used to query for it

File: src/rdkafka_cgrp.c [hunk 13]
```diff
@@ -6462,11 +6462,27 @@ void rd_kafka_cgrp_metadata_update_check(rd_kafka_cgrp_t *rkcg,
...
+        if (rkcg->rkcg_group_protocol != RD_KAFKA_GROUP_PROTOCOL_GENERIC)
+                return;
+
         if (!rkcg->rkcg_subscription || rkcg->rkcg_subscription->cnt == 0)
                 return;
 
-        if (rkcg->rkcg_group_protocol != RD_KAFKA_GROUP_PROTOCOL_GENERIC)
+        if (!(rkcg->rkcg_flags & RD_KAFKA_CGRP_F_WILDCARD_SUBSCRIPTION) &&
+            rkcg->rkcg_join_state != RD_KAFKA_CGRP_JOIN_STATE_WAIT_METADATA &&
+            rkcg->rkcg_join_state != RD_KAFKA_CGRP_JOIN_STATE_STEADY) {
```
[the remark this answers is not in the document]
Comment (emasab) [reply src/rdkafka_cgrp.c:6473]: For sure we know that in states RD_KAFKA_CGRP_JOIN_STATE_WAIT_INIT and RD_KAFKA_CGRP_JOIN_STATE_WAIT_JOIN it can cause the issue #4589 . I was also considering to put `rkcg_join_state >= RD_KAFKA_CGRP_JOIN_STATE_WAIT_METADATA` but doing these updates with a stable group seems more error proof.
In RD_KAFKA_CGRP_JOIN_STATE_WAIT_METADATA it's required because there are cases where it's waiting exactly this update.
For a regex subscription it's required in state INIT too, because it has to find the topics to include in the JoinGroup request, but in that case it doesn't have the problem with metadata cache.

**Comment by emasab** [comment]
Closing in favour of

#4676 
#4677
#4678
#4679
#4680
